Add Tag Manager datastore

aaemnnosttv opened the pull request:
## Summary

Addresses issue #1385 

## Relevant technical choices

* Added [`@jackfranklin/test-data-bot`](https://www.npmjs.com/package/@jackfranklin/test-data-bot) (formerly `test-data-bot`) for generating test data in unit tests (similar to the `factory()` we have in PHP unit tests).
    - Used for generating random data for accounts and container objects rather than relying on fixtures for everything
* Added factory function for html with/without an existing tag for use in tests with mocked tag verification requests
* Updated `GET:tag-permission` datapoint to always return a successful response, with a response format similar to Analytics, returning `{ accountID: string, containerID: string, permission: bool }`
* Added `POST:create-container` datapoint
* Added `GET:live-container-version` datapoint
* Removed unnecessary `POST:settings` datapoint (route handled in `Modules` now and only handles get/update settings)
* Renamed `isAmp` selector on `core/site` to `isAMP` for consistency with our naming conventions


## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/accounts.js:
```diff
@@ -0,0 +1,189 @@
+		const registry = yield Data.commonActions.getRegistry();
+		registry.dispatch( STORE_NAME ).setAccountID( accountID );
+
+		if ( ACCOUNT_CREATE === accountID ) {
+
+		}
+
+		// Trigger cascading selections.
```
This part is from the same action in Analytics that we will likely need here as well. Will implement this part in #1386.

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/accounts.js:
```diff
@@ -0,0 +1,190 @@
+		const { error } = state;
+
+		return error || null;
+	},
```
Done! Maybe we should add something like this to `commonStore` or `createModuleStore` instead? I took the same file from AdSense for consistency and only had to update doc blocks.

felixarntz replied:
I don't think we should do that for now, since we're still not clear whether this is the error handling we'll wanna have in the long term. While I agree this is bad for DRY, let's not abstract it for this reason just yet.

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/containers.js:
```diff
@@ -0,0 +1,150 @@
+		return { accountID };
+	},
+	controlCallback: ( { accountID } ) => {
+		return API.get( 'modules', 'tagmanager', 'containers', { accountID }, { useCache: false } );
```
For some reason I thought it was already like that on the server, but you're right. Updated 👍

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/existing-tag.js:
```diff
@@ -0,0 +1,191 @@
+			containerID: tag,
+			permission,
+		}, { tag } );
+	},
```
I mostly added these to help with the migration. Will remove.

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/existing-tag.js:
```diff
@@ -0,0 +1,191 @@
+			...state,
+			tagPermission: {
+				...state.tagPermission,
+				[ tag ]: permission,
```
I agree that there is value in the association that comes back from this as it is useful for a subsequent containers request, but I'm not sure that storing the account ID with the permission itself is the right place for it.

If the intention is to store references to which container belongs to which account, that should probably have its own selector and state which maps `[ containerID ]: accountID`. It doesn't make sense to check tag permission to get a related accountID. Is there another reason why it should be stored here?

felixarntz replied:
Since the account ID is being detected as part of the permission check, it's stored here. I agree it feels a bit odd, but this is how we've been doing it in the other areas as well. I'd prefer to stick to this convention for now, also to get this in, and discuss via a new issue how we can improve that.

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/existing-tag.js:
```diff
@@ -0,0 +1,191 @@
+	 * @return {(boolean|undefined)} Permission
+	 */
+	hasTagPermission( state, tag ) {
+		const permission = state.tagPermission[ tag ];
```
No problem to add this if needed; see my response to your comment there 🙂

aaemnnosttv commented on assets/js/modules/tagmanager/util/validation.js:
```diff
@@ -0,0 +1,98 @@
+ * @param {(string|number)} internalContainerID Internal container ID to test.
+ * @return {boolean} Whether or not the given ID is valid.
+ */
+export { isValidNumericID as isValidInternalContainerID };
```
We do the same in Analytics which is what this was based on.

Would you be okay with doing it this way instead?
```js
export const isValidInternalContainerID = ( internalContainerID ) => isValidNumericID( internalContainerID );
```

aaemnnosttv replied:
Update - I rewrote them as functions as there were already a few of these in the file 😄

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/existing-tag.js:
```diff
@@ -0,0 +1,191 @@
+
+const fetchGetTagPermissionStore = createFetchStore( {
+	baseName: 'getTagPermission',
+	argsToParams: ( tag ) => {
```
Yeah, I agree it's less ambiguous and have updated it accordingly. The only downside is that there is potential confusion with `containerID` (the setting) which is web-container specific. In this case `containerID` is a public ID for any found container. Maybe we should call it `existingContainerID`?

I've updated it to use `containerID` for now and also added support for this on the server side as well.

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/accounts.js:
```diff
@@ -0,0 +1,190 @@
+		registry.dispatch( STORE_NAME ).setAccountID( accountID );
+
+		if ( ACCOUNT_CREATE === accountID ) {
+
```
My editor is stripping the return there because it's an eslint violation (unnecessary return). Surprisingly it doesn't complain about the empty conditional 😆 

Will add when adding the cascading selections there.

aaemnnosttv commented:
Thanks for all the feedback @felixarntz - I think I've addressed all of it now (including those places where it wasn't commented on specifically).

The only thing left to clarify there is around the tag permission changes. See my reply to your comment above.

aaemnnosttv commented:
@felixarntz this should be ready for another pass. I cleaned up the tests a bit as well and introduced a new [`untilResolved`](https://github.com/google/site-kit-wp/pull/1676/commits/69e257680911e9b066e8f520529b54b23144fc4c) utility which really cleans up the usage of `hasFinishedResolution` - I think you'll really like it 😄

tofumatt commented on assets/js/modules/tagmanager/util/validation.js:
```diff
@@ -0,0 +1,116 @@
+ * @return {boolean} Whether or not the given container ID is valid.
+ */
+export function isValidContainerID( containerID ) {
+	return !! containerID?.toString?.()?.match?.( /^GTM-[A-Z0-9]+$/ );
```
These `?` are making me feel like I'm writing Swift. I kinda dig it! I wasn't aware optional chaining was so well-supported and old! Just an ES feature I wasn't in the habit of using, but it's slick. 👍

tofumatt replied:
That said, the `?.()?` syntax is a bit tough at first! But I get it, and it's cool. 🙂

aaemnnosttv replied:
Yeah, this is a bit of an extreme example, but it really simplifies the implementation for longer calls. See https://github.com/google/site-kit-wp/commit/3d2508cd2160adbe49b9195a6f54e7e6cc19498b

tofumatt commented on tests/js/utils.js:
```diff
@@ -135,6 +137,34 @@ export const subscribeWithUnsubscribe = ( registry, ...args ) => {
+ * @return {Proxy} Proxy instance that dynamically returns a function to await a selector's
+ *                 resolution by the same name, with the same arguments.
+ */
+export const untilResolved = ( registry, storeName ) => {
```
![very nice I like](https://user-images.githubusercontent.com/90871/85434139-824f5c80-b57d-11ea-942d-0462469c6f6a.jpg)

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/containers.js:
```diff
@@ -0,0 +1,171 @@
+	},
+	controlCallback: ( { accountID } ) => {
+		// Always request both contexts to prevent filtering on server.
+		// TODO: Remove `usageContext` param when legacy component is removed and datapoint
```
No issue specifically for this that I'm aware of, but it can be done as part of #1386 which will remove the legacy component as part of it 👍

aaemnnosttv commented on assets/js/modules/tagmanager/datastore/containers.js:
```diff
@@ -0,0 +1,171 @@
+	 * @return {boolean} True if a request for create-container is in progress, otherwise false.
+	 */
+	isDoingCreateContainer( state ) {
+		return Object.values( state.isFetchingCreateContainer ).some( Boolean );
```
Maybe, but it's not the only place we're doing it, so this is consistent with other methods that do the same thing.

felixarntz commented on assets/js/modules/tagmanager/datastore/containers.js:
```diff
@@ -0,0 +1,171 @@
+	 * @param {string} usageContext Container usage context. (Either 'web', or 'amp')
+	 * @return {Object} Object with `response` and `error`.
+	 */
+	*createContainer( accountID, usageContext ) {
```
Along the lines of what we decided in #716, I wonder whether we should make `usageContext` property of an object? We could keep it required or alternatively default to `web`. This diverges from the ACs, but wanted to put it up for consideration.

aaemnnosttv replied:
I like this idea although we should be careful about using defaults in the datastore due to the high possibility of `undefined` being passed for something. It's not unreasonable to think that we might allow for the container name to be customized at creation time - it's actually more important than a profile name since you can't have two containers with the same name, but there are almost no restrictions on view names.

felixarntz replied:
Sounds good, let's keep it required for now, but move into an object since it could easily become one of _multiple_ creation arguments.

felixarntz commented on assets/js/modules/tagmanager/datastore/containers.js:
```diff
@@ -0,0 +1,171 @@
+};
+
+const baseResolvers = {
+	*getContainers( accountID ) {
```
Does this work? The resolver here has different parameters from the selector.

aaemnnosttv replied:
Yep - the resolver doesn't need to have the same signature as the selector, it is called with the same arguments though (less `state` of course).

felixarntz commented on assets/js/modules/tagmanager/datastore/containers.js:
```diff
@@ -0,0 +1,171 @@
+	 * @param {string} [usageContext] Usage context of containers to filter by.
+	 * @return {(Array|undefined)} Array of containers, or `undefined` if not loaded yet.
+	 */
+	getContainers( state, accountID, usageContext ) {
```
See my other comment, we may want to put `usageContext` into an object here. It's an optional filter argument, so I think that would make sense. Or we could remove this parameter from here entirely, may make more sense, and then introduce `getWebContainers` and `getAMPContainers` selectors which would rely on `getContainers` and then filter accordingly.

aaemnnosttv replied:
I like the idea of the dedicated context-specific selectors 👍
